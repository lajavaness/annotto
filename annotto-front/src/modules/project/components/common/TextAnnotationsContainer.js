import { CheckCircleFilled } from '@ant-design/icons'
import { Input } from 'antd'
import { debounce, isEmpty } from 'lodash'
import { useTranslation } from 'react-i18next'
import PropTypes from 'prop-types'
import React, { useCallback, useEffect, useRef, useState } from 'react'

import theme from '__theme__'

import * as Styled from './__styles__/TextAnnotationsContainer.styles'

const TextAnnotationsContainer = ({ tasks, annotations, onChange, showPredictions, predictions }) => {
	const rootRef = useRef()

	const [textAnnotations, setTextAnnotations] = useState([])

	const { t } = useTranslation('project')

	const resolvePrediction = useCallback(
		(value) => (!isEmpty(predictions) ? predictions.find((e) => e.value === value) : []),
		[predictions]
	)

	const resolveTextAnnotation = useCallback(
		(value) => textAnnotations.find((e) => e.value === value),
		[textAnnotations]
	)

	const updateAnnotations = useCallback(
		debounce((newAnnotations) => {
			return !!onChange && onChange(newAnnotations)
		}, 50),
		[onChange]
	)

	const updateTextAnnotations = useCallback(
		(taskValue, text) => {
			const filteredAnnotation = !isEmpty(annotations) ? annotations.filter(({ value }) => value !== taskValue) : []

			const newAnnotations = text ? [...filteredAnnotation, { value: taskValue, text }] : filteredAnnotation

			setTextAnnotations(newAnnotations)
			return updateAnnotations(newAnnotations)
		},
		[annotations, updateAnnotations]
	)

	useEffect(() => {
		setTextAnnotations(!isEmpty(annotations) ? annotations : [])
	}, [annotations])

	const _onChange = useCallback(
		(taskValue) =>
			({ target: { value: text } }) =>
				updateTextAnnotations(taskValue, text),
		[updateTextAnnotations]
	)

	const _onValidateClick = useCallback(
		(taskValue, text) => () => updateTextAnnotations(taskValue, text),
		[updateTextAnnotations]
	)

	return (
		<Styled.Root ref={rootRef}>
			{tasks.map(({ label, value }, index) => (
				<Styled.Container key={`${index}_${value}`}>
					<Styled.LabelContainer>
						{!isEmpty(predictions) && showPredictions && resolvePrediction(value) && (
							<Styled.PredictionIcon
								data-testid={`predictionIcon-${index}`}
								twoToneColor={theme.colors.predictionIconPrimary}
							/>
						)}
						<Styled.Label>{label}</Styled.Label>
					</Styled.LabelContainer>
					<Styled.TextAreaContainer>
						<Input.TextArea
							data-testid={`textAnnotationArea-${index}`}
							row={3}
							placeholder={
								(showPredictions && resolvePrediction(value)?.text) ||
								t('project:annotation.textAnnotation.placeholder')
							}
							allowClear
							onChange={_onChange(value)}
							value={resolveTextAnnotation(value)?.text}
						/>
						{showPredictions && !!resolvePrediction(value)?.text && !resolveTextAnnotation(value)?.value && (
							<Styled.ValidateButton
								type="ghost"
								onClick={_onValidateClick(value, resolvePrediction(value)?.text)}
								icon={<CheckCircleFilled />}
							/>
						)}
					</Styled.TextAreaContainer>
				</Styled.Container>
			))}
		</Styled.Root>
	)
}

export default TextAnnotationsContainer

const TaskValue = PropTypes.string

const TaskShape = PropTypes.shape({
	/** A machine-readable key that identifies the label in the backend, the value is unique.*/
	value: TaskValue.isRequired,
	/** The text displayed in the list for annotators to recognise. */
	label: PropTypes.string.isRequired,
})

TextAnnotationsContainer.propTypes = {
	/** Defines the tasks to display. */
	tasks: PropTypes.arrayOf(TaskShape),
	/** The list of value that are currently annotated. Callees must manage this
	 * list in their state. */
	annotations: PropTypes.arrayOf(
		PropTypes.shape({
			/** A machine-readable key that identifies the task value of an annotation in the backend,
			 * the value is unique. */
			value: TaskValue.isRequired,
			/** The text of an annotation. Displayed in the corresponding input. */
			text: PropTypes.string,
		})
	).isRequired,
	/** Define if predictions should appear on labels. */
	showPredictions: PropTypes.bool,
	/** Defines the values of predictions. */
	predictions: PropTypes.arrayOf(
		PropTypes.shape({
			/** A machine-readable key that identifies the task value of an annotation in the backend,
			 * the value is unique. */
			value: TaskValue.isRequired,
			/** The text of an annotation. Displayed in the corresponding input. */
			text: PropTypes.string,
		})
	),
	/** A callback that will be called whenever a label is annotated or unannotated.
	 * Contains the updated list of annotated ids as the first parameter. */
	onChange: PropTypes.func,
}

TextAnnotationsContainer.defaultProps = {
	tasks: [],
	annotations: [],
	onChange: null,
	showPredictions: true,
	predictions: [],
}
