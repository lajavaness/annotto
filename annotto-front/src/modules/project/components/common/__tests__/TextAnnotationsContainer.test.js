import { ThemeProvider } from 'styled-components'
import { cleanup, fireEvent, render } from '@testing-library/react'
import React from 'react'

import theme from '__theme__'

import TextAnnotationsContainer from 'modules/project/components/common/TextAnnotationsContainer'

const defaultProps = {}
const getInstance = (props = {}) => (
	<ThemeProvider theme={theme}>
		<TextAnnotationsContainer {...defaultProps} {...props} />
	</ThemeProvider>
)

const tasks = [
	{
		value: 'foo',
		label: 'foo',
	},
	{
		value: 'bar',
		label: 'bar',
	},
]
const annotations = [
	{
		value: 'foo',
		text: 'foo_text',
	},
	{
		value: 'bar',
		text: 'bar_text',
	},
]
const showPredictions = true
const predictions = [
	{
		value: 'foo',
		text: 'foo_text',
	},
	{
		value: 'bar',
		text: 'bar_text',
	},
]
const onChange = jest.fn()

describe('TextAnnotationsContainer', () => {
	jest.useFakeTimers()
	afterEach(cleanup)

	it('matches snapshot', () => {
		const { asFragment } = render(getInstance())
		expect(asFragment()).toMatchSnapshot()
	})

	it('matches snapshot with props', () => {
		const { asFragment } = render(getInstance({ tasks, annotations, onChange, showPredictions, predictions }))
		expect(asFragment()).toMatchSnapshot()
	})

	it('render labels of tasks', () => {
		const { getByText } = render(getInstance({ tasks }))
		tasks.forEach(({ label }) => expect(getByText(label)).toBeVisible())
	})

	it('render annotations', () => {
		const { getByText } = render(getInstance({ tasks, annotations }))
		annotations.forEach(({ text }) => expect(getByText(text)).toBeVisible())
	})

	it('render predictions when no annotations', () => {
		const { getByPlaceholderText } = render(getInstance({ tasks, predictions }))
		predictions.forEach(({ text }) => expect(getByPlaceholderText(text)).toBeVisible())
	})

	it('render predictionIcon if prediction and show prediction', () => {
		const { getByTestId } = render(getInstance({ tasks, predictions, showPredictions }))
		predictions.forEach((prediction, index) => expect(getByTestId(`predictionIcon-${index}`)).toBeVisible())
	})

	it('triggers onChange callback', () => {
		const { getByTestId } = render(getInstance({ tasks, annotations, onChange }))

		fireEvent.change(getByTestId(`textAnnotationArea-${0}`), { target: { value: 'zog' } })

		expect(onChange).not.toHaveBeenCalled()

		jest.runAllTimers()

		expect(onChange).toBeCalledTimes(1)
		expect(onChange).toHaveBeenCalledWith([
			{
				value: 'bar',
				text: 'bar_text',
			},
			{ value: 'foo', text: 'zog' },
		])
	})
})
