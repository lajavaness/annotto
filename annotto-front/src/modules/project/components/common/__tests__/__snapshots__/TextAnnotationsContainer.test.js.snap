// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextAnnotationsContainer matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gsBqHK iPExTj"
  />
</DocumentFragment>
`;

exports[`TextAnnotationsContainer matches snapshot with props 1`] = `
<DocumentFragment>
  <div
    class="sc-gsBqHK iPExTj"
  >
    <div
      class="sc-aKXVD eryilG"
    >
      <div
        class="sc-cvJIAo gpnNIA"
      >
        <span
          aria-label="tags"
          class="anticon anticon-tags sc-gIRjGK kULWEc"
          data-testid="predictionIcon-0"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="tags"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M477.5 694l311.9-311.8-19-224.6-224.6-19-311.9 311.9L477.5 694zm116-415.5a47.81 47.81 0 0133.9-33.9c16.6-4.4 34.2.3 46.4 12.4a47.93 47.93 0 0112.4 46.4 47.81 47.81 0 01-33.9 33.9c-16.6 4.4-34.2-.3-46.4-12.4a48.3 48.3 0 01-12.4-46.4z"
              fill="#fff2f0"
            />
            <path
              d="M476.6 792.6c-1.7-.2-3.4-1-4.7-2.3L137.7 456.1a8.03 8.03 0 010-11.3L515.9 66.6c1.2-1.3 2.9-2.1 4.7-2.3h-.4c-2.3-.2-4.7.6-6.3 2.3L135.7 444.8a8.03 8.03 0 000 11.3l334.2 334.2c1.8 1.9 4.3 2.6 6.7 2.3z"
              fill="#fff2f0"
            />
            <path
              d="M889.7 539.8l-39.6-39.5a8.03 8.03 0 00-11.3 0l-362 361.3-237.6-237a8.03 8.03 0 00-11.3 0l-39.6 39.5a8.03 8.03 0 000 11.3l243.2 242.8 39.6 39.5c3.1 3.1 8.2 3.1 11.3 0l407.3-406.6c3.1-3.1 3.1-8.2 0-11.3zM652.3 337.3a47.81 47.81 0 0033.9-33.9c4.4-16.6-.3-34.2-12.4-46.4a47.93 47.93 0 00-46.4-12.4 47.81 47.81 0 00-33.9 33.9c-4.4 16.6.3 34.2 12.4 46.4a48.3 48.3 0 0046.4 12.4z"
              fill="#ed5857"
            />
            <path
              d="M137.7 444.8a8.03 8.03 0 000 11.3l334.2 334.2c1.3 1.3 2.9 2.1 4.7 2.3 2.4.3 4.8-.5 6.6-2.3L861.4 412c1.7-1.7 2.5-4 2.3-6.3l-25.5-301.4c-.7-7.8-6.8-13.9-14.6-14.6L522.2 64.3h-1.6c-1.8.2-3.4 1-4.7 2.3L137.7 444.8zm408.1-306.2l224.6 19 19 224.6L477.5 694 233.9 450.5l311.9-311.9z"
              fill="#ed5857"
            />
          </svg>
        </span>
        <span
          class="ant-typography sc-kmATbt"
        >
          foo
        </span>
      </div>
      <div
        class="sc-dkmKIT hZeDhN"
      >
        <span
          class="ant-input-affix-wrapper ant-input-affix-wrapper-textarea-with-clear-btn"
        >
          <textarea
            class="ant-input"
            data-testid="textAnnotationArea-0"
            placeholder="foo_text"
            row="3"
          >
            foo_text
          </textarea>
          <span
            aria-label="close-circle"
            class="anticon anticon-close-circle ant-input-clear-icon"
            role="button"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              data-icon="close-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
    <div
      class="sc-aKXVD eryilG"
    >
      <div
        class="sc-cvJIAo gpnNIA"
      >
        <span
          aria-label="tags"
          class="anticon anticon-tags sc-gIRjGK kULWEc"
          data-testid="predictionIcon-1"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="tags"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M477.5 694l311.9-311.8-19-224.6-224.6-19-311.9 311.9L477.5 694zm116-415.5a47.81 47.81 0 0133.9-33.9c16.6-4.4 34.2.3 46.4 12.4a47.93 47.93 0 0112.4 46.4 47.81 47.81 0 01-33.9 33.9c-16.6 4.4-34.2-.3-46.4-12.4a48.3 48.3 0 01-12.4-46.4z"
              fill="#fff2f0"
            />
            <path
              d="M476.6 792.6c-1.7-.2-3.4-1-4.7-2.3L137.7 456.1a8.03 8.03 0 010-11.3L515.9 66.6c1.2-1.3 2.9-2.1 4.7-2.3h-.4c-2.3-.2-4.7.6-6.3 2.3L135.7 444.8a8.03 8.03 0 000 11.3l334.2 334.2c1.8 1.9 4.3 2.6 6.7 2.3z"
              fill="#fff2f0"
            />
            <path
              d="M889.7 539.8l-39.6-39.5a8.03 8.03 0 00-11.3 0l-362 361.3-237.6-237a8.03 8.03 0 00-11.3 0l-39.6 39.5a8.03 8.03 0 000 11.3l243.2 242.8 39.6 39.5c3.1 3.1 8.2 3.1 11.3 0l407.3-406.6c3.1-3.1 3.1-8.2 0-11.3zM652.3 337.3a47.81 47.81 0 0033.9-33.9c4.4-16.6-.3-34.2-12.4-46.4a47.93 47.93 0 00-46.4-12.4 47.81 47.81 0 00-33.9 33.9c-4.4 16.6.3 34.2 12.4 46.4a48.3 48.3 0 0046.4 12.4z"
              fill="#ed5857"
            />
            <path
              d="M137.7 444.8a8.03 8.03 0 000 11.3l334.2 334.2c1.3 1.3 2.9 2.1 4.7 2.3 2.4.3 4.8-.5 6.6-2.3L861.4 412c1.7-1.7 2.5-4 2.3-6.3l-25.5-301.4c-.7-7.8-6.8-13.9-14.6-14.6L522.2 64.3h-1.6c-1.8.2-3.4 1-4.7 2.3L137.7 444.8zm408.1-306.2l224.6 19 19 224.6L477.5 694 233.9 450.5l311.9-311.9z"
              fill="#ed5857"
            />
          </svg>
        </span>
        <span
          class="ant-typography sc-kmATbt"
        >
          bar
        </span>
      </div>
      <div
        class="sc-dkmKIT hZeDhN"
      >
        <span
          class="ant-input-affix-wrapper ant-input-affix-wrapper-textarea-with-clear-btn"
        >
          <textarea
            class="ant-input"
            data-testid="textAnnotationArea-1"
            placeholder="bar_text"
            row="3"
          >
            bar_text
          </textarea>
          <span
            aria-label="close-circle"
            class="anticon anticon-close-circle ant-input-clear-icon"
            role="button"
            tabindex="-1"
          >
            <svg
              aria-hidden="true"
              data-icon="close-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
              />
            </svg>
          </span>
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
